task: prefetch: Condense the RHSM entitlement RPM repositories logic to a single step

eskultety opened the pull request:
~**Depends on:**  https://github.com/containerbuildsystem/cachi2/pull/792~

Opening as an RFC draft since this solution currently relies on a custom-built cachi2 image which the task was tested with in a private Konflux deployment. Once the dependent PR is merged AND this solution is accepted I will lift the Draft status. Pasting a (publishable) snippet from the pipelines log:
```
prefetch
=======
Registering with Red Hat subscription manager.
The system has been registered with ID: cc44f731-1d42-44f2-9993-56454f63faa4
The registered system name is: konflux-test-repo-on-pull-rdc2a0834b9fb80d8cdb850c8f3586779-pod
2025-01-22 17:02:50,177 INFO Fetching the gomod dependencies at subpath hello
2025-01-22 17:02:50,186 INFO go.mod reported versions: '1.20'[go], '-'[toolchain]
2025-01-22 17:02:50,187 INFO Using Go release: go1.20
2025-01-22 17:02:50,195 INFO Downloading the gomod dependencies
2025-01-22 17:02:52,891 INFO Retrieving the list of packages
2025-01-22 17:02:53,195 INFO Reading RPM lockfile: /var/workdir/source/rpms.lock.yaml
2025-01-22 17:02:53,197 INFO Downloading files for 'x86_64' architecture.
2025-01-22 17:02:53,203 INFO Using client certificate auth.
2025-01-22 17:02:58,373 INFO All dependencies fetched successfully \o/
2025-01-22 17:02:59,454 INFO Creating repository metadata for repoid 'appstream': /var/workdir/cachi2/output/deps/rpm/x86_64/appstream
2025-01-22 17:02:59,520 INFO Creating /var/workdir/cachi2/output/deps/rpm/x86_64/repos.d/cachi2.repo
Unregistering from: subscription.rhsm.redhat.com:443/subscription
System has been unregistered.

build
====
...
[2/2] STEP 7/8: CMD ["/usr/local/bin/hello"]
[2/2] STEP 8/8: LABEL "build-date"="2025-01-22T17:03:57" "architecture"="x86_64" "vcs-type"="git" "vcs-ref"="a0c881286921d0ae047b932d3dc1a7645394c122" "quay.expires-after"="5d"
[2/2] COMMIT quay.io/<path>/konflux-test-repo:on-pr-a0c881286921d0ae047b932d3dc1a7645394c122
--> 60f0fc490100
Successfully tagged quay.io/<path>/konflux-test-repo:on-pr-a0c881286921d0ae047b932d3dc1a7645394c122
60f0fc49010053f7915bd33d1c646d2015832fc51161576a87ea33ce5851e9ba
```

eskultety commented:
Worth noting that although early failure isn't currently enforced in Bash, the solution as proposed should still employ something like `trap 'subscription-manager unregister' EXIT`.

MartinBasti commented on task/prefetch-dependencies-oci-ta/0.1/prefetch-dependencies-oci-ta.yaml:
```diff
@@ -325,6 +183,50 @@ spec:
       script: |
         #!/bin/bash
 
```
Can we use this opt `set -euo pipefail` to make sure that we fail early, and then use `trap` to unregister? It's confusing to users to see script continues when there was a failures. It generates misleading output.

eskultety replied:
I thought it was proposed in a different PR, but sure, I can enable it in a separate commit and as I mentioned in a different response to the PR *add the trap*.

MartinBasti replied:
it was, but this PR changed the task completelly

eskultety commented on task/prefetch-dependencies-oci-ta/0.1/prefetch-dependencies-oci-ta.yaml:
```diff
@@ -381,6 +273,35 @@ spec:
+        if [ -e /activation-key/org ]; then
+          RHSM_ORG=$(cat /activation-key/org)
+          RHSM_ACT_KEY=$(cat /activation-key/activationkey)
+          creds="$(cat /activation-key/org /activation-key/activationkey)"
+          IFS=$'\n' read -r RHSM_ORG RHSM_ACT_KEY <<<"$creds"
```
These 2 lines should not be here, that's some (clearly) failed rebase attempt, consider them deleted in favour of the 2 lines above them.

eskultety commented:
Since last revision:

- enforce early failure with `set -euo pipefail`
- add a trap handler for RHSM unregister

Both were tested in a private deployment, attaching the prefetch task log:
```
step-prefetch-dependencies
INFO: Using mounted CA bundle: /mnt/trusted-ca/ca-bundle.crt
'/mnt/trusted-ca/ca-bundle.crt' -> '/etc/pki/ca-trust/source/anchors/ca-bundle.crt'
Registering with Red Hat subscription manager.
The system has been registered with ID: 97eac960-858d-4de6-b641-bebd57b15c04
...
Unregistering from: subscription.rhsm.redhat.com:443/subscription
System has been unregistered.

step-create-trusted-artifact
2025/01/24 12:29:08 Skipping step because a previous step failed
```

chmeliik commented on task/prefetch-dependencies-oci-ta/0.1/prefetch-dependencies-oci-ta.yaml:
```diff
@@ -381,6 +277,33 @@ spec:
+          RHSM_ACT_KEY=$(cat /activation-key/activationkey)
+
+          echo "Registering with Red Hat subscription manager."
+          subscription-manager register \
+            --org "${RHSM_ORG}" \
+            --activationkey "${RHSM_ACT_KEY}" || exit 1
+
```
IMO the cleanest way to ensure unregistration would be to `trap 'subscription-manager unregister' EXIT` just below the register call

No need to check the existence of `/activation-key/org` in the trap, because we'd only be setting up the trap in cases where it's really needed. And no need to handle unregistration again at the end of the task script

eskultety replied:
> IMO the cleanest way to ensure unregistration would be to trap 'subscription-manager unregister' EXIT just below the register call

That is cleaner, then again, unregister itself can fail which would make the whole task fail. Unregistering is a good practice, but I think it should be best effort. I mean I wouldn't want to force users to rerun pipelines just because there's some intermittent error on the unregister endpoint, so I'd like to keep it as a function. I'll drop the check as you suggest and move the trap call right after the register sequence.

chmeliik replied:
:+1: 

For what it's worth, I think that traps don't change the overall exit status unless you explicitly call `exit N` in the trap itself. But it's best to be explicit about it, one never knows with Bash. (+1 to keep the function)

chmeliik commented on task/prefetch-dependencies-oci-ta/0.1/prefetch-dependencies-oci-ta.yaml:
```diff
@@ -325,10 +183,58 @@ spec:
+                      }'
+            )"
+          elif (jq -e . 2>/dev/null 1>&2 <<<"$input"); then
+            input_json_has_rpm="$(jq -e '
```
Careful with `jq -e` - in combination with `set -e`, it will cause the script to exit if the expression evaluates to something falsy

chmeliik commented on task/prefetch-dependencies-oci-ta/0.1/prefetch-dependencies-oci-ta.yaml:
```diff
@@ -325,10 +183,58 @@ spec:
+            )"
+          elif (jq -e . 2>/dev/null 1>&2 <<<"$input"); then
+            input_json_has_rpm="$(jq -e '
+                                    if type == "array" then
+                                      any(.[]; .type == "rpm")
+                                    else
+                                      .type == "rpm"
```
Cachi2 also recognizes this form of input I think

```json
{
  "packages": [{"type": "rpm"}],
  "flags": []
}
```

eskultety replied:
Ouch, correct! And just like that the `jq` part will turn into a write-only code segment (getting reminiscent of Perl) :) .

chmeliik reviewed (commented):
Looks like a better approach :+1:

eskultety commented:
Since last revision:

- dropped the 'jq **-e**' usage
- incorporated handling of the '{"packages": []}' input JSON form
- formatting (mostly indentation) changes

chmeliik commented on task/prefetch-dependencies-oci-ta/0.1/prefetch-dependencies-oci-ta.yaml:
```diff
@@ -336,8 +339,10 @@ spec:
 
-        # UNREGISTER RHSM
-        rhsm_unregister
+        # hack: the OCI generator would delete the function since it doesn't consider trap a "usage"
+        if false; then
+          rhsm_unregister
+        fi
```
Lol, that's a funny one. Can't think of a better solution, so LGTM :smile:

chmeliik reviewed (commented):
LGTM once the image is updated to the official cachi2 image

MartinBasti reviewed (approved):
LGTM with note about updating image before merging

eskultety commented:
Update: we're facing CI issues on cachi2 side (https://github.com/containerbuildsystem/cachi2/pull/789/checks?check_run_id=36275813585) preventing us from rolling the corresponding release out which blocks this PR from having the Draft status lifted and the image references updated.

eskultety commented:
All blockers resolved, this one should be good to go.

chmeliik commented:
/test build-definitions-pull-request

chmeliik commented:
Found this in e2e-tests logs
```
      pod: test-comp-hzmi-on-pull-request-8xvb5-prefetch-dependencies-pod | container step-prefetch-dependencies: 
      2025-01-30 08:40:07,538 ERROR InvalidInput: 1 validation error for user input\npackages -> 0\n  Input tag '' found using 'type' does not match any of the expected tags: 'bundler', 'generic', 'gomod', 'npm', 'pip', 'rpm', 'yarn'
      Error: InvalidInput: 1 validation error for user input
      packages -> 0
        Input tag '' found using 'type' does not match any of the expected tags: 'bundler', 'generic', 'gomod', 'npm', 'pip', 'rpm', 'yarn'
```

chmeliik commented:
> Found this in e2e-tests logs
> 
> ```
>       pod: test-comp-hzmi-on-pull-request-8xvb5-prefetch-dependencies-pod | container step-prefetch-dependencies: 
>       2025-01-30 08:40:07,538 ERROR InvalidInput: 1 validation error for user input\npackages -> 0\n  Input tag '' found using 'type' does not match any of the expected tags: 'bundler', 'generic', 'gomod', 'npm', 'pip', 'rpm', 'yarn'
>       Error: InvalidInput: 1 validation error for user input
>       packages -> 0
>         Input tag '' found using 'type' does not match any of the expected tags: 'bundler', 'generic', 'gomod', 'npm', 'pip', 'rpm', 'yarn'
> ```

It was for this repo at this commit I think https://github.com/konflux-qe-bd/devfile-sample-python-basic/blob/b1240032adb024160ba7a76e5fc5fbeef19cc4c0/.tekton/test-comp-hzmi-pull-request.yaml

It doesn't look like it's setting prefetch-input, so it most likely defaulted to empty string :thinking:

chmeliik commented on task/prefetch-dependencies/0.1/prefetch-dependencies.yaml:
```diff
@@ -82,164 +82,13 @@ spec:
-    - name: INPUT
-      value: $(params.input)
-    script: |
-      if [ -z "${INPUT}" ]
-      then
-        # Confirm input was provided though it's likely the whole task would be skipped if it wasn't
-        echo "No prefetch will be performed because no input was provided for cachi2 fetch-deps"
-        echo "skip" > /shared/skip
-      fi
```
I think we need to replicate this skip logic somewhere

eskultety replied:
Sigh, great catch! I reverted this bit of logic prior to commit 5f62fe3b01edcd6897a00c1c156d4f7272a0e31e .

chmeliik commented:
/retest

eskultety commented:
> @chmeliik: The following test has Completed, say **/retest** to rerun failed tests.
> PipelineRun Name 	Status 	Rerun command 	Build Log 	Test Log
> `build-definitions-pull-request-8r2np` 	**Completed** 	`/retest` 	[View Pipeline Log](https://console-openshift-console.apps.stone-prd-rh01.pg1f.p1.openshiftapps.com/k8s/ns/konflux-ci/tekton.dev~v1~PipelineRun/build-definitions-pull-request-8r2np) 	[View Test Logs](https://console-openshift-console.apps.stone-prd-rh01.pg1f.p1.openshiftapps.com/k8s/ns/konflux-ci/tekton.dev~v1~PipelineRun/build-definitions-pull-request-8r2np/logs?taskName=e2e-tests)
> ### Inspecting Test Artifacts
> 
> To inspect your test artifacts, follow these steps:
> 
>     1. Install **ORAS** (see the [ORAS installation guide](https://oras.land/docs/installation)).
> 
>     2. Download artifacts with the following commands:
> 
> 
> ```shell
> mkdir -p oras-artifacts
> cd oras-artifacts
> oras pull quay.io/konflux-test-storage/konflux-team/build-definitions:5240b65fb94ba188f6f6e7e7ac3b9fb2d30e259e
> ```
> 
> ### Test results analysis
> 
> 🚨 **Error occurred while running the E2E tests, list of failed Spec(s)**:
> 
> ➡️ [**`failed`**] [It] [build-service-suite Build templates E2E test] HACBS pipelines build-definitions ec pipelines runs ec pipeline pipelines/enterprise-contract.yaml [build, build-templates, HACBS, pipeline, build-templates-e2e]
> Click to view logs

Is this a legitimate failure related to the changes or an infra problem?

chmeliik commented:
> Is this a legitimate failure related to the changes or an infra problem?

Infra I think

/retest

chmeliik commented:
> > > Is this a legitimate failure related to the changes or an infra problem?
> > 
> > 
> > Infra I think
> 
> I haven't seen other PRs failing the same way. Is it necessary to pass for a PR to be accepted? I mean, sure it's totally preferred, do we need to report this anywhere?

If critical, someone with admin permissions can bypass the merge requirements (e.g. Mykola, Andrew). But generally tests do need to pass.

I've seen this happen at least twice before, and it happened twice here as well. It's intermittent, but seems common. So yeah, I'll raise the problem
